### PR TITLE
Dm 2361

### PR DESCRIPTION
**Pull request by JoshingYou1**
### JIRA issue link
https://agile6.atlassian.net/browse/DM-2361

## Description - what does this code do?
Adds search functionality and facilities table to VISN show page.

## Testing done - how did you test it/steps on how can another person can test it 
_**Search Feature**_
1. Add/update any number of practices' origin facilities at least one practices' initiating facility to facilties that belong to a single VISN.
2. visit the show page of that VISN.
3. scroll down the the 'Search for Practices' section and make sure the `practices created in this visn` radio is selected (default choice based on design's mockup).
4. Make sure the amount of initial results matches the intro blurb created count at the top of the page.
5. Click on the `practices adopted in this VISN` radio input and make sure the resulting count matches the intro blurb adopted count at the top of the page.
6. Add text to the search field and hit enter.
7. Make sure a query string is added to the URL and search results get updated appropriately.
8. Clear the search bar and click on the categories combo box.
9. Take note of the possible category options for the adopted practices.
10. Now click on the `created` radio again and check the combo box. If the created practices have more/less categories, you should see a difference. The combo box is populated based on which radio input is selected.
11. Get back to the state where you have the initial results for the created practices.
12. Use the sorting options as you see fit to make sure they are working as intended.
13. To test the load more feature, make sure you have at least 7 results for a given radio input (created or adopted).
14. Now add some text to the search field again and run the search.
15. Make sure there's at least one result and click on a practice card.
16. Once you arrive to that practice's show page, make sure the VISN number breadcrumb also has the query string you created when you ran the search (ex: `/visns/1?query=test`).
17. Click on the breadcrumb link and make sure you return the VISN show page with the query string present and the results that correspond with that search query.
18. Add a bunch of random characters to the search field and run the search.
19. Make sure you seen the no results icon and text.
### UPDATE 5/3/21
20. Open the inspection tool in your browser and go to the `console` tab.
21. Make sure the only times the results get updated (the form is submitted) are the following:
- The value of the combo box has changed.
- The radio input has changed.
- The search field value has changed.
22. Find a VISN in your DB that does not have any adopted practices or created practices, but does have the other type (ex: 0 adopted, 3 created). Visit that VISN's show page.
23. Make sure the radio input, on initial page load, is defaulted to whichever type is populated with at least one practice (ex: If there are 2 adopted practices and 0 created practices, make sure the adopted practices radio input is selected).
24. In `visns/show.html.erb`, add this code:
- Line 20: visn_practices_created_count = 0
- Line 22: visn_practices_adopted_count = 0
25. Refresh the page and make sure the search portion of the page is no longer there. After talking with UX, we decided to not show the search portion of the page if there are neither created nor adopted practices for a particular VISN.

_**Tracking Search**_
1. Open up pgAdmin or whatever tool you use to look at the DB collections.
2. Run a search and make sure you see a new `ahoy_event` called `VISN search`. Also make sure the text matches your query.
### UPDATE 5/3/21
Play with the combo box and radio inputs in order to make sure an `ahoy_event` is created ONLY when the search field is populated.

_**Facilities Table**_
1. Scroll down below the search section to the facilities table.
2. Test out the sorting feature for all of the columns and make sure it works as intended.
3. Click on the question mark icon in the the `Complexity` column and make sure a modal appears with text describing each complexity level. Click on the `X` at the top right to close the modal.
4. Click on a facility of your choosing and make sure you're directed to the show page of that facility.

_**BONUS (Loading Spinner)**_
The search is very fast naturally, but if you want to see the spinners, you can add this block of code to several areas:
```
setTimeout(function() {
   // The code goes here
}, 2000)
```
Wrap the above code around the following code:
1. Around line `295` in `visns/_show.js.erb`, or you can search for the text `Run the search`.
2. Around line `298` in visns/_show.js.erb, or you can search for the text `Sort the results based on each sort option`. Add it to each case (`buildSearchResults` functions).
3. Around line `332` in `visns/_show.js.erb`, or you can search for the code `search(decodeURI(query))`;
4. Around line `168` in `visns/_show.js.erb`, or you can search for the text `Get the results and default to filtering based on most adoptions` (`buildSearchResults` function).
5. Around line `90` in showMoreItems.js, or you can search for the code `for (var i = _.nowNum; i < _.goNum; i++) {`. Include both the For loop and the truncateText() function.
```
for (var i = _.nowNum; i < _.goNum; i++) {
                $(element).children().eq(i).show()
                _.nowNum += 1
                $('.addListData').closest('section').find(loadingSpinner).addClass('display-none');
                $(buttonBox).removeClass('display-none');
                // If there are no more practice cards to show, hide the 'No more results' button
                if (_.nowNum === _.sum) {
                    $(buttonBox).addClass('display-none');
                }
            }
            // Truncate the text of the next set of cards
            truncateText();
```

## Screenshots, Gifs, Videos from application (if applicable)
Screen Shot 2021-04-29 at 11 40 28 AM
Screen Shot 2021-04-29 at 11 40 38 AM

## Link to mock-ups/mock ups (image file if you have it) (if applicable)
https://www.figma.com/file/fNXjZzLUS39bKmjdfDWyxy/My-VISN%2FVAMC

## Acceptance criteria
- Each VISN page should have “Practices Created” and “Practices Adopted” radio button, defaults to Practices Created
- Each VISN should have search for practice within VISN 
- Category dropdown search
- Sort dropdown filtering (most adoptions, recently created, and A to Z) 
- 6 results to start then load more
- List of VAMCs in VISN at bottom of page

## Definition of done
- [X] Unit tests written (if applicable)
- [X] e2e/accessibility tests written (if applicable)
- [X] Events are logged appropriately
- [ ] Documentation has been updated, if applicable
- [X] A link has been provided to the originating JIRA issue
- [X] No sensitive information (i.e. PII/credentials/internal URLs/etc.) is captured in logging, hardcoded, or specs

### CODE REVIEW

File: app/controllers/visns_controller.rb [hunk 1]
```diff
@@ -54,6 +54,36 @@ def show
...
     end
+
+    searchable_practices = Practice.searchable_practices nil
+    # set '@practices_json' to avoid js console error when utilizing the practices/search.js.erb file
+    @practices_json = practices_json(searchable_practices)
+
+    practices_created_by_visn = []
+    helpers.get_created_practices_by_visn(searchable_practices, @visn, practices_created_by_visn)
+
+    def get_categories_by_practices(practices, practice_categories)
```
[the remark this answers is not in the document]
Comment (JoshingYou1) [reply app/controllers/visns_controller.rb:65]: done!!

File: app/views/visns/_show.js.erb [hunk 2]
```diff
@@ -0,0 +1,373 @@
...
+        if (query !== undefined) {
+            searchField.value = decodeURI(query);
+        } else {
+            searchField.value = '';
+        }
+    }
+}
+
+function trackVisnSearchField(searchTerm) {
+    if (searchTerm !== '') {
```
[the remark this answers is not in the document]
Comment (JoshingYou1) [reply app/views/visns/_show.js.erb:338]: Fixed! Thank you for your assistance on this 😄

**Review by bradjohnson92008 — COMMENTED** [review]
Great Job on this Josh...  I was able to verify all the functionality - and the feature looks great!   Only thing I noticed is that it looks like there is a syntax error on the resultsForOnlyRadioVal() method... (but I debugged the code and it works fine.. :)  I also saw some console.log calls in that file as well.  _show.js.erb.

**Comment by JoshingYou1** [comment]
I'm not seeing any errors for that method on my local
Screen Shot 2021-05-04 at 7 46 13 PM

Console logs have been removed!

File: app/views/visns/_show.js.erb (hunk 2 above)
[the remark this answers is not in the document]
Comment (emyl3) [reply app/views/visns/_show.js.erb:335]: Thank you for addressing this! Great work on this brain teaser hahah 🧠

**Review by bradjohnson92008 — APPROVED** [review]
Great job Josh!  LGTM!

**Pull request merged**